src/main/python/mmlspark/cyber/anomaly/collaborative_filtering.py

```diff
@@ -0,0 +1,880 @@
+__author__ = 'rolevin'
+
+from typing import List, Optional, Tuple
+
+from mmlspark.cyber.anomaly.complement_access import ComplementAccessTransformer
+from mmlspark.cyber.feature import indexers, scalers
+from mmlspark.cyber.utils import spark_utils
+
+import numpy as np
+
+from pyspark.ml import Estimator, Transformer
+from pyspark.ml.param.shared import Param, Params
+from pyspark.ml.recommendation import ALS
+from pyspark.sql import DataFrame, functions as f, types as t
+
+"""
+Glossary:
+the term 'res' in this package is a shorthand for resource
+"""
+
+
+def _make_dot():
+    """
+    create a method that performs a dot product between two vectors (list of doubles)
+    :return: the method
+    """
+    @f.udf(t.DoubleType())
+    def dot(v, u):
+        if (v is not None) and (u is not None):
+            vv = np.pad(np.array(v), (0, len(u) - len(v)), 'constant', constant_values=1.0) if len(v) < len(
+                u) else np.array(v)
+            uu = np.pad(np.array(u), (0, len(v) - len(u)), 'constant', constant_values=1.0) if len(u) < len(
+                v) else np.array(u)
+
+            return float(vv.dot(uu))
+        else:
+            return None
+
+    return dot
+
+
+class CfAlgoParams:
```

Comment:
This class should use sparkml parameters and should be mixed into the classes that need it

Reply:
Will remove CfAlgoParams

Reply:
removed as agreed